cvmfs/sqlitevfs.cc

```diff
@@ -0,0 +1,494 @@
+/**
+ * This file is part of the CernVM File System.
+ *
+ * An optimized virtual file system layer for the client only.  It expects to
+ * operate on immutable, valid SQlite files.  Hence it can do a few
+ * optimiziations.  Most notably it doesn't need to know about the path of
+ * the SQlite file once opened.  It works purely on the file descriptor.
+ */
+
+#include "cvmfs_config.h"
+#include "sqlitevfs.h"
+
+#include <dlfcn.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdint.h>
+#include <sys/select.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <sys/uio.h>
+#include <unistd.h>
+
+#include <cassert>
+#include <climits>
+#include <cstring>
+#include <ctime>
+
+#include "duplex_sqlite3.h"
+#include "platform.h"
+#include "smalloc.h"
+#include "statistics.h"
+
+using namespace std;  // NOLINT
+
+namespace sqlite {
+
+namespace {
+
+const char *kVfsName = "cvmfs-readonly";
+
+/**
+ * The private user data attached to the sqlite_vfs object.
+ */
+struct VfsRdOnly {
+  VfsRdOnly()
+    : n_access(NULL)
+    , no_open(NULL)
+    , n_rand(NULL)
+    , sz_rand(NULL)
+    , n_read(NULL)
+    , sz_read(NULL)
+    , n_sleep(NULL)
+    , sz_sleep(NULL)
+    , n_time(NULL)
+  { }
+  perf::Counter *n_access;
+  perf::Counter *no_open;
+  perf::Counter *n_rand;
+  perf::Counter *sz_rand;
+  perf::Counter *n_read;
+  perf::Counter *sz_read;
+  perf::Counter *n_sleep;
+  perf::Counter *sz_sleep;
+  perf::Counter *n_time;
+};
+
+/**
+ * This is passed to all operations once a file is opened.
+ */
+struct VfsRdOnlyFile {
+  sqlite3_file base;  // Base class. Must be first.
+  VfsRdOnly *vfs_rdonly;
+  int fd;
+  uint64_t size;
+};
+
+}  // anonymous namespace
+
+
+static int VfsRdOnlyClose(sqlite3_file *pFile) {
+  VfsRdOnlyFile *p = reinterpret_cast<VfsRdOnlyFile *>(pFile);
+  int retval = close(p->fd);
+  if (retval == 0) {
+    perf::Dec(p->vfs_rdonly->no_open);
+    return SQLITE_OK;
+  }
+  return SQLITE_IOERR_CLOSE;
+}
+
+
+/**
+ * On a short read, the remaining bytes must be zero'ed.
+ * TODO(jblomer): the reads seem to be rather small.  Investigate buffered read.
+ */
+static int VfsRdOnlyRead(
+  sqlite3_file *pFile,
+  void *zBuf,
+  int iAmt,
+  sqlite_int64 iOfst
+) {
+  VfsRdOnlyFile *p = reinterpret_cast<VfsRdOnlyFile *>(pFile);
+  ssize_t got = pread(p->fd, zBuf, iAmt, iOfst);
```

Comment:
Perhaps.  Borderline because they connect to sqlite and reflect the variable names from the sqlite documentation.  We have a similar situation with the calling conventions in tbb.  Please feel free to change if you think it makes the code more readable.